episodes/03-muscle-connect.md

```diff
@@ -0,0 +1,418 @@
+:::::
+
+:::
+
+## The constructor
+
+To let a model communicate with the muscle manager, other submodels and the outside world we need to create a `libmuscle.Instance` object:
+
+```python
+from libmuscle import Instance
+from ymmsl import Operator
+
+def some_example_submodel():
+    instance = Instance({
+            Operator.F_INIT: ['initial_state_a', 'initial_state_b'],
+            Operator.O_I: ['some_intermediate_state'],
+            Operator.S: ['some_other_intermediate_state'],
+            Operator.O_F: ['final_state_a', 'final_state_b']})
+```
+
+The constructor takes a single argument, a dictionary that maps `ymmsl.Operator` objects to lists of ports. Ports are used by submodels to exchange messages with the outside world and eachother. These messages usually contain the state of the model, but they can also contain other information. Ports can be input ports or output ports, but not both at the same time, the operator that they are mapped to determines how the port is used. The names of the available operators correspond to their theoretical counterparts ($F_{init}$, $O_I$, $S$ and $O_F$ ).
+
+::: challenge
+
+## What ports do we need?
+
+Question: In our example of the reaction submodel, what ports do we need to communicate the model state to the outside world? Come up with good names. What operators would we need to map them to?
+
+:::::: solution
+
+In this model, we need:
+
+- a single input port that will receive an initial state (e.g. named "initial_state") at the beginning of the model run and it should be mapped to the $F_{init}$ operator
+- a single output port that sends the final state (e.g. named "final_State") at the end of the reaction simulation to the rest of the simulation, it should be mapped to the $O_F$ operator
+::::::
+
+:::
+
+In multiscale simulations, submodels often have to run multiple times, for instance because they are used as a micro model or because they are part of an ensemble that cannot be completely parallelised. In order to accomplish this, the entire submodel is wrapped into a loop. Exactly when this loop needs to end often depends on the behaviour of the whole model, and is not easy to determine beforehand, but fortunately MUSCLE will do that for us if we call the `libmuscle.Instance.reuse_instance()` method.
+
+```python
+    while instance.reuse_instance():
+        # F_INIT
+        # State update loop
+        # O_F
+```
+
+::: challenge
+
+## Exercise 2: Constructing your submodel
+
+Add the following to our reaction model code:
+
+- a `libmuscle.Instance` object with the appropriate operator-port dictionary
+- the reuse loop
+
+:::: solution
+
+```python
```

Comment:
"Exercise 2: Constructing your submodel" suggests (to me) that you should have a working/finished submodel at the end of this exercise. I would therefore probably attempt to replace the `F_INIT` section with code that gets the values from `initial_state` and then get stuck. Is there a way we can clarify here that the "solution" will only be a work in progress, with further steps needed to finish the submodel?

Reply:
I solved this by renaming the exercise, let me know if you think this is enough.